File: cpp/test/user_entity_test.cpp
// Generated basic-flow test for the user entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct UserSetup {
  std::shared_ptr<BluefinShieldconexMgmtSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static UserSetup user_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/user/UserTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinShieldconexMgmtSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("user01"), Value("user02"), Value("user03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID", idmap},
    {"BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value("FALSE")},
    {"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_USER_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value("TRUE");

  UserSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void user_entity_instance() {
  auto testsdk = BluefinShieldconexMgmtSDK::testSDK();
  auto ent = testsdk->user();
  ASSERT_EQ(ent->getName(), std::string("user"), "entity name");
}

static void user_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"user", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinShieldconexMgmtSDK::testSDK(seed, sdkopts);
  auto se = strsdk->user();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinShieldconexMgmtSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->user();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void user_entity_basic() {
  auto setup = user_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"load"}) {
    auto sk = is_control_skipped("entityOp", std::string("user.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;

  // Bootstrap entity data from existing test data (no create step in flow).
  // Declare _data at FUNCTION scope (later load/update steps reference it);
  // only _data_raw was declared, so the block-local assignment left _data
  // undeclared ("was not declared in this scope").
  Value user_ref01_data_raw = Helpers::toMapAny(Struct::getpath(setup.data, {"existing", "user"}));
  Value user_ref01_data = vmap();
  {
    std::vector<Value> its = Struct::items(user_ref01_data_raw);
    user_ref01_data = its.empty() ? vmap() : Helpers::toMapAny(pair_val(its[0]));
    if (!user_ref01_data.is_map()) user_ref01_data = vmap();
  }
  // LOAD
  auto user_ref01_ent = client->user();
  Value user_ref01_match_dt0 = vmap({{"id", getp(user_ref01_data, "id")}});
  Value user_ref01_data_dt0_loaded = user_ref01_ent->load(Struct::clone(user_ref01_match_dt0), Value::undef());
  Value user_ref01_data_dt0_load_result = Helpers::toMapAny(user_ref01_data_dt0_loaded);
  ASSERT_TRUE(user_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(user_ref01_data_dt0_load_result, "id"), getp(user_ref01_data, "id"), "expected load result id to match");

}

int main() {
  T_RUN(user_entity_instance);
  T_RUN(user_entity_stream);
  T_RUN(user_entity_basic);
  return sdktest::summary("user_entity_test");
}

File: scala/sdktest/UserEntityTest.scala
// Generated basic-flow test for the user entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped UserTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefinshieldconexmgmtsdk.core.{Helpers, BluefinShieldconexMgmtSDK}
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct

object UserEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("user.instance") {
      val testsdk = BluefinShieldconexMgmtSDK.testSDK()
      val ent = testsdk.user(null)
      rep.check("user.instance", ent != null, "expected non-null user entity")
    }

    rep.scope("user.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/user/UserTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinShieldconexMgmtSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("user01", "USER01")
      idmap.put("user02", "USER02")
      idmap.put("user03", "USER03")
      val now = System.currentTimeMillis()

      // LOAD
      val userRef01Ent = client.user(null)
      val userRef01MatchDt0 = new LinkedHashMap[String, Object]()
      userRef01MatchDt0.put("id", userRef01Data.get("id"))
      val userRef01DataDt0Loaded = userRef01Ent.load(userRef01MatchDt0, null)
      val userRef01DataDt0LoadResult = Helpers.toMapAny(userRef01DataDt0Loaded)
      rep.check("user.load.map", userRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("user.load.id", userRef01Data.get("id"), userRef01DataDt0LoadResult.get("id"))
    }
  }
}
